### ISSUE TITLE
Sparsely blocked/chunked arrays

### ISSUE THREAD

**Issue by system123** [dask/dask#7652]
I have an application whereby I am manually forming a dask array using `blocks` as each of the chunks in the final array need to go through a specific transform to ensure the correct alignment in the final dask array. Furthermore, not all the chunks in the array which is formed contain useful information, but some are just needed to ensure spatial alignment between their neighboring chunks, thus forming a Dask array with sparse blocks/chunks.

Currently I am using masked arrays to mask out entire chunks, however, this is inefficient as those blocks are still included in the computational graph and are still evaluated to some degree. Ultimately it would be nice to have the concept of chunk/block masks which let dask know that the evaluation of any operation on an entire chunk will result in a NaN or empty result and can thus be skipped entirely without additional computation. 

The application generally has more `empty` blocks/chunks that those with valid data and thus the computational graph grows very large with a lot of unneeded operations, having a way to prevent this would be nice.

A small example of how I currently handle this:
```
import numpy as np
import dask.array as da

arr0 = da.from_array(np.arange(1, 26).reshape(5,5), chunks=(5, 5))
arr1 = da.from_array(np.arange(25, 50).reshape(5,5), chunks=(5, 5))
# This chunk is empty/blank/invalid but is needed to align the other chunks
arr2 = da.from_array(np.full((5,5), np.nan), chunks=(5, 5)) 
arr3 = da.from_array(np.arange(75, 100).reshape(5,5), chunks=(5, 5))

a = da.block([[arr0, arr1],[arr2, arr3]])
b = da.ma.masked_invalid(a)
c = b.min().compute()
```

What would be nice to have:
```
import numpy as np
import dask.array as da

arr0 = da.from_array(np.arange(1, 26).reshape(5,5), chunks=(5, 5))
arr1 = da.from_array(np.arange(25, 50).reshape(5,5), chunks=(5, 5))
arr3 = da.from_array(np.arange(75, 100).reshape(5,5), chunks=(5, 5))

a = da.sparse_block([[arr0, arr1],[None, arr3]])
b = a.min().compute()
```

**Comment by GenevieveBuckley** [comment]
Perhaps what's needed here is better culling of tasks for masked arrays?

**Comment by system123** [comment]
That would probably be the easier approach, just check if the entire chunk is masked and if so then cull it.

**Comment by jsignell** [comment]
This seems vaguely similar to your work in #7655 right @GenevieveBuckley?

**Comment by GenevieveBuckley** [comment]
I don't think they're very similar. Right now https://github.com/dask/dask/pull/7655 is targeted only at low level slicing, not fancy indexing or masking. I'll see what Ian & Gabe think about it when I talk to them next.

**Comment by SultanOrazbayev** [comment]
I would like to express interest in this feature also. To give more context, I am interested in performing some operations on large sparse arrays, however given the scale of the matrices the number of tasks grows excessively even though most of those tasks are useless (e.g. adding a block of zeros to another block of zeros).

One strategy is removing tasks from `.dask` graph based on the chunks/blocks that are known to contain zeros, but there might be a better way of dealing with this and avoid creating tasks that are known to be unnecessary.

Do you think it's feasible to add a block-level boolean attribute to designate if the block is `empty` (or all zeros), e.g. `self.block_is_empty` (True/False/None for empty, non-empty, unknown)? Then, the status of each block can be used to reduce tasks or pre-compute them (e.g. element-wise multiplication of an empty block with a non-empty block is an empty block). If there is interest in this, I can work on a PR with some guidance.

**Comment by SultanOrazbayev** [comment]
This is a small example, but with some awareness about emptiness in the graph below only 3 tasks would be needed:

```python
from dask.array import eye

I = eye(3, chunks=1)

print(f"Number of tasks:", len((2 * I).dask))
# Number of tasks: 18
```

**Comment by jsignell** [comment]
> Do you think it's feasible to add a block-level boolean attribute to designate if the block is `empty` (or all zeros), e.g. `self.block_is_empty` (True/False/None for empty, non-empty, unknown)?

This feels equivalent to `sum(self.block.shape) == 0`, if you agree then I think the challenge is less how to tell which chunks are empty and more how to pluck those chunks out of the array and get rid of any dependencies on them.

**Comment by SultanOrazbayev** [comment]
I'm not sure I understand this approach correctly... shape of a block containing zeros is non-zero, right?

**Comment by jsignell** [comment]
I think the approach that you have makes sense and if you plan to put it at the block-level then I think you won't get complaints. So it would be like `self.blocks[0,1].is_empty`?

**Comment by SultanOrazbayev** [comment]
Yes, except given your comment, perhaps a better name is `.is_all_zeros` (or something that emphasizes zeros). The reason `zeros` is good to emphasize is because the implemented logic should be special to the case of zero-filled matrix (so if someone had a `sparse` matrix with non-zero fill value, the optimizations would not necessarily apply).

Some of the optimizations possible for blocks that contain only zero values:
- no need to multiply with the other block (both elementwise and dot), can return block of zeros (with exceptions if the other block has nan/inf)
- no need to add with the other block, can return the other block
- no need to transpose
- can apply ufunc to a single zero block and re-use the result for all zero blocks (maybe this is too much optimization)

There's probably more... for very large sparse arrays, this can lead to a substantial reduction in computational graph (these computations are fast, it's more a question of not burdening the scheduler), I hope however this doesn't too add much overhead for dense arrays with few zero-only blocks.

**Comment by system123** [comment]
I'd suggest `is_nodata` or something along those lines to abstract it further away from any specific value (for instance, in my application, these values are often `nan`). 

To avoid any impact to dense arrays, perhaps we want to subclass this into a `sparse_array` so that there is no performance hit for standard arrays (which I'd assume are a large percentage of use cases)

**Comment by hansmohrmann** [comment]
I am also registering my interest in such a feature! I have been trying both by using the `sparse ` library (per [this ](https://docs.dask.org/en/latest/array-sparse.html)reference) and by using `da.masked_array`, but it does not appear that either reduces computation in the case of very sparse array operations (even just the example on the `array-sparse` page above runs slower on using `sparse`).

As for `is_zero `vs `is_nodata`, I am not sure these are entirely equivalent; in the case of elementwise multiplication, for instance, optimizing over an entire block of `zeros `is perfectly legitimate, and the result is mathematically zero; if that block is instead all `nan`, the desired result might be `nan `instead of `zero`, so I'm not sure if it should be generalized from zero to nan.

**Comment by SultanOrazbayev** [comment]
Thanks, @hansmohrmann ! I got in touch with @system123 and promised to tag him once I get a meaningful progress on a PR. I can tag you also, or if you want to work on it in sync (or any other action plan) kindly get in touch by email or LinkedIn and we can coordinate on a call.

**Comment by system123** [comment]
@hansmohrmann fair point. I think taking that into account we could design this to optimize both for the case where all values are `zero` as this continues to hold mathematically, as well as for the case were all values are `masked` as the mask is essentially telling us that there are no guarentees about the data points underneath. This way in a case such as mine I can convert my `nan` points to a data mask (as I tried before) and work from there.

**Comment by ljstrnadiii** [comment]
It's been a while, but I’d like to express interest in this feature and share a couple of use cases.

For example, I might want to build a large mosaic of all tiles intersecting the coast of Australia using rioxarray with a VRT or GTI. Most of the interior in such a case would be filled with NaN or nodata. Another common scenario is building an array that spans a continent, but only contains data for ~5000 scattered fields of interest. In that case, I usually need to maintain an index for all 5000 arrays. Most of this would happen through Xarray and zarr, but task graph implications are relevant to this thread. 

I’d also love support for map_overlap_blocks, and the ability to infer a mask that nullifies blocks corresponding to missing Zarr chunks. This would allow workflows where I skip writing empty chunks but can still use a chunk-level validity mask to index the array efficiently.
